src/basic/hashmap.c

```diff
@@ -1841,3 +1870,101 @@ int set_put_strsplit(Set *s, const char *v, const char *separators, ExtractFlags
                         return r;
         }
 }
+
+/* expand the cachemem if needed, return true if newly (re)activated. */
+static int cachemem_maintain(CacheMem *mem, unsigned size) {
+        int r = false;
+        const void **p;
+
+        assert(mem);
+
+        if (size > mem->n_allocated) {
+                /* TODO: grow by chunks at a time */
+                p = realloc(mem->ptr, size * sizeof(*mem->ptr));
```

Comment:
GREEDY_REALLOC() appears to be a good candidate here?

Reply:
Yep, I'm not accustomed to all the convenience helper stuff in systemd.